ITE/10/README.adoc

```diff
@@ -0,0 +1,278 @@
+```
+
+Each `expectedStepPredicate` object has the following schema.
+
+```yaml
+predicateType: string
+expectedProducts: list of artifact rules
+functionaries: list of authorized functionary key IDs
+threshold: int
+```
+
+The `predicateType` field identifies the type of predicate expected for the
+step. `expectedProducts` is moved to being predicate specific as informational
+predicates may be generated alongside transformational predicates for the same
+step. This means that product rules do not generalize across all of the step's
+predicates. However, `expectedMaterials` is retained at the step level as all
+predicates generated for a step are expected to be for the same set of
+artifacts. `pubkeys` is updated to `functionaries` to generalize signing
+semantics. Finally, `threshold` applies per predicate type rather than the step
+as a whole.
+
+Similarly, inspections are also updated to support generating different
+predicate types.
+
+```yaml
+name: string
+command: list of strings
+expectedMaterials: list of artifact rules
+expectedPredicates: list of expectedInspectionPredicates
+```
+
+Each `expectedInspectionPredicate` object has the following schema.
+
+```yaml
+predicateType: string
+expectedProducts: list of artifact rules
+```
+
+As the attestations generated during inspection executions are not signed,
+`functionaries` and `threshold` are omitted.
+
+=== Supporting Sublayouts
+
+The in-toto specification has support for sublayouts which are steps in a layout
+that have not links but a bundle of another layout and links which are
+recursively verified first. To support this semantic, this ITE proposes making
+sublayout declarations explicit. Therefore, a step declaration declares that it
+expects a layout and identifies it. This layout is then verified recursively
+before verification continues at the original layout.
+
+=== Verifying Artifact Rules
+
+The verification workflow in the in-toto specification matches links to use by
+the name of the step and the key ID used to sign it. This part of the workflow
+is updated to support matching on predicate types as well.
+
+```go
+func Verify(layoutEnvelope, layoutKeys, attestations, now) {
+    verifyLayoutSignatures(layoutEnvelope, layoutKeys)
+
+    layout := layoutEnvelope.Payload
+    verifyLayoutExpiry(layout, now)
+
+    if layout.Steps == nil && layout.Inspections == nil {
+        panic
+    }
+
+    for _, step := range layout.Steps {
+        stepAttestations := attestationsForStep(step.Name, attestations)
+        for _, predicate := range step.ExpectedPredicates {
+            predicateAttestations := attestationsForStepPredicate(predicate.Type, stepAttestations)
+            if predicate.Type == LAYOUT {
+                Verify(predicateAttestations[0], predicate.Functionaries, predicateAttestations[1:], now)
+            } else {
+                verifyAttestationSignatures(predicate.Functionaries, predicate.Threshold, predicateAttestations)
```

Comment:
Will this error of a signature doesn't match or will it just remove unverified attestations from the list?

I suspect it would be better if it just filtered `predicateAttestations` to only have attestations that were verified and allow verification to continue if some signatures can't be verified.

Rationale: Perhaps new attestations get added to the bundle that aren't yet trusted by the layout, but also aren't necessary.

Reply:
This is an interesting observation.  We should consider the sort of problem you raise as a more general case.  

For example, I can also see situations where an organization wants to use a stricter sublayout, perhaps in a way that says that all dependencies must match some criteria.  So, certainly there, having additional information that isn't needed, shouldn't cause verification to fail.

Reply:
I'll update the workflow pseudocode but the idea is that all attestations for each step have their signatures verified. The ones that pass sig verification are retained after which a threshold check is applied. So attestations from untrusted keys will not abort verification if the threshold is met. This is also in-toto v1 behaviour.